### PR TITLE
hardening/1152_configurable_http_parameters

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #1152 
* 100% unit tests passed:
* Relevant tests for this PR:

`NGSIHDFSSink` tests:
```
[NGSIHDFSSink.configure] -------------------------------------------- When not configured, not mandatory parameters get default values
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'backend.impl=rest' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'backend.max_conns=500' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'backend.max_conns_per_route=100' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'batch_size=1' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'batch_timeout=30' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'batch_ttl=30' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'csv_separator=,' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'data_model=dm-by-entity' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'enable_encoding=false' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'enable_grouping=false' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'enable_lowercase=false' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'file_format=json-row' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'hdfs_host=localhost' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'hdfs_port=14000' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'enable_hive=false' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'enable_krb5=false' configured by default
[NGSIHDFSSink.configure] -------------------------------------  OK  - 'service_as_namespace=false' configured by default
[NGSIHDFSSinkTest.configure] ---------------------------------------- backend.max_conns gets the configured value
[NGSIHDFSSinkTest.configure] ---------------------------------  OK  - 'backend.max_conns=25' was configured
NGSIHDFSSinkTest.configure] ---------------------------------------- backend.max_conns_per_route gets the configured value
[NGSIHDFSSinkTest.configure] ---------------------------------  OK  - 'backend.max_conns_per_route=3' was configured
```
`NGSICKANSink` tests:
```
[NGSICKANSink.configure] -------------------------------------------- When not configured, not mandatory parameters get default values
[NGSICKANSink.configure] -------------------------------------  OK  - 'attr_persisnce=row' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'api_key=nokey' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'backend.max_conns=500' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'backend.max_conns_per_rpute=100' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'batch_size=1' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'batch_timeout=30' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'batch_ttl=30' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'data_model=dm-by-entity' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'enable_encoding=false' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'enable_grouping=false' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'enable_lowercase=true' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'ckan_host=localhost' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'ckan_port=80' configured by default
[NGSICKANSink.configure] -------------------------------------  OK  - 'ssl=false' configured by default
NGSICKANSink.configure] -------------------------------------------- backend.max_conns gets the configured value
[NGSICKANSink.configure] -------------------------------------  OK  - 'backend.max_conns=25' was configured
[NGSICKANSink.configure] -------------------------------------------- backend.max_conns_per_route gets the configured value
[NGSICKANSink.configure] -------------------------------------  OK  - 'backend.max_conns_per_route=3'
```
`NGSICartoDBSink` tests:
```
[NGSICartoDBSink.configure] ----------------------------------------- When not configured, not mandatory parameters get default values
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'backend.max_conns=500' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'backend.max_conns_per_route=100' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'batch_size=1' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'batch_timeout=30' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'batch_ttl=30' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'data_model=dm-by-entity' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'enable_distance=false' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'enable_grouping=false' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'enable_lowercase=true' configured by default
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'enable_raw=true' configured by default
[NGSICartoDBSink.configure] ----------------------------------------- backend.max_conns gets the configured value
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'backend.max_conns=25' was configured
[NGSICartoDBSink.configure] ----------------------------------------- backend.max_conns gets the configured value
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'backend.max_conns=25' was configured
[NGSICartoDBSink.configure] ----------------------------------------- backend.max_conns_per_route gets the configured value
[NGSICartoDBSink.configure] ----------------------------------  OK  - 'backend.max_conns_per_route=3' was configured
```

### CODE REVIEW

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java [hunk 1]
```diff
@@ -27,28 +29,132 @@
...
+        String region = null;
+        String secretAccessKey = "my_secret_access_key";
+        NGSIDynamoDBSink sink = new NGSIDynamoDBSink();
+        sink.configure(createContext(accessKeyId, attrPersistence, batchSize, batchTimeout, batchTTL, dataModel,
+                enableGrouping, enableLowercase, region, secretAccessKey));
+        
+        try {
+            assertTrue(sink.getRowAttrPersistence());
+            System.out.println(getTestTraceHead("[NGSIDynamoDBSink.configure]")
+                    + "-  OK  - 'attr_persisnce=row' configured by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java]: attr_persisnce=row -> attr_persistence=row

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java [hunk 2]
```diff
@@ -27,28 +29,132 @@
...
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSIDynamoDBSink.configure]")
+                    + "- FAIL - 'attr_persisnce=row' not configured by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java]: 'attr_persisnce=row' -> attr_persistence=row

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java [hunk 3]
```diff
@@ -43,13 +44,182 @@ public NGSICKANSinkTest() {
...
+        String port = null; // default
+        String ssl = null; // default
+        NGSICKANSink sink = new NGSICKANSink();
+        sink.configure(createContext(apiKey, attrPersistence, backendMaxConns, backendMaxConnsPerRoute, batchSize,
+                batchTime, batchTTL, dataModel, enableEncoding, enableGrouping, enableLowercase, host, port, ssl));
+        
+        try {
+            assertTrue(sink.getRowAttrPersistence());
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "-  OK  - 'attr_persisnce=row' configured by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: 'attr_persisnce=row' -> 'attr_persistence=row'

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java [hunk 4]
```diff
@@ -43,13 +44,182 @@ public NGSICKANSinkTest() {
...
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "- FAIL - 'attr_persisnce=row' not configured by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: 'attr_persisnce=row' -> 'attr_persistence=row'

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java [hunk 5]
```diff
@@ -43,13 +44,182 @@ public NGSICKANSinkTest() {
...
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "- FAIL - 'backend.max_conns=500' not configured by default");
+            throw e;
+        } // try catch
+        
+        try {
+            assertEquals(100, sink.getBackendMaxConnsPerRoute());
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "-  OK  - 'backend.max_conns_per_rpute=100' configured by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: 'backend.max_conns_per_rpute=100' -> 'backend.max_conns_per_route=100'

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java [hunk 6]
```diff
@@ -43,13 +44,182 @@ public NGSICKANSinkTest() {
...
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "- FAIL - 'backend.max_conns_per_rpute=100' not configured by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: 'backend.max_conns_per_rpute=100' -> 'backend.max_conns_per_route=100'

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 7]
```diff
@@ -41,13 +43,219 @@ public NGSIHDFSSinkTest() {
...
+            throw e;
+        } // try catch
+        
+        try {
+            assertTrue(!sink.getEnableLowerCase());
+            System.out.println(getTestTraceHead("[NGSIHDFSSink.configure]")
+                    + "-  OK  - 'enable_lowercase=false' configured by default");
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSIHDFSSink.configure]")
+                    + "- FAIL - 'enable_owercase=false' not configured by default");
```
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java [hunk 8]
```diff
@@ -56,24 +56,156 @@ public NGSICartoDBSinkTest() {
...
+            throw e;
+        } // try catch
+        
+        try {
+            assertTrue(!sink.getEnableDistance());
+            System.out.println(getTestTraceHead("[NGSICartoDBSink.configure]")
+                    + "-  OK  - 'enable_distance=false' configured by default");
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSICartoDBSink.configure]")
+                    + "- FAIL - 'enable_encoding=false' not configured by default");
```
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java [hunk 9]
```diff
@@ -43,13 +44,182 @@ public NGSICKANSinkTest() {
...
+            throw e;
+        } // try catch
+        
+        try {
+            assertTrue(sink.getEnableLowerCase());
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "-  OK  - 'enable_lowercase=true' configured by default");
+        } catch (AssertionError e) {
+            System.out.println(getTestTraceHead("[NGSICKANSink.configure]")
+                    + "- FAIL - 'enable_owercase=true' not configured by default");
```
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java (hunk 6 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java (hunk 5 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java (hunk 4 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java (hunk 3 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICKANSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java (hunk 2 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java (hunk 1 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIDynamoDBSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1162/commits/e8604df05d9f9d9033f2810069d8c8cc76633587

**Pull request merged**